### ISSUE TITLE
I/O failures when using encrypted disks

### ISSUE THREAD

**Issue by pykello** [ubicloud/ubicloud#276]
If I create 3 VMs with encrypted disks, and do `sudo apt update`, then `sudo apt install gcc`, then `sudo apt upgrade` inside each VM, I get the following error:

```
update-initramfs: Generating /boot/initrd.img-5.15.0-75-generic
sync: error syncing '/boot/initrd.img-5.15.0-75-generic': Input/output error
```

Or sometimes `Read-only filesystem` error.

When I look at SPDK logs using `journalctl -u spdk.service`, I see 150 of the following errors.

```
vhost[1069]: bdev_aio.c: 202:bdev_aio_writev: *ERROR*: bdev_aio_writev: io_submit returned -14
spdk[1069]: bdev_aio.c: 202:bdev_aio_writev: *ERROR*: bdev_aio_writev: io_submit returned -14
```

I don't see them in the unencrypted mode.

When I look at output of `strace -p [spdk pid] 2>&1 | grep io_submit`, I see 75 lines with `EFAULT` error codes. In unencrypted mode, I see 0 `EFAULT` errors.

```
io_submit(0x7fadb0fc3000, 1, [{aio_data=0x20006d799920, aio_lio_opcode=IOCB_CMD_PWRITEV, aio_fildes=36, aio_buf=[{iov_base=0x8000000000000000, iov_len=65536}], aio_offset=3043033088}]) = -1 EFAULT (Bad address)
```

Also in VM serial logs we see errors like:

```
[   51.534463] Buffer I/O error on device vda1, logical block 439673
...
[   51.556688] blk_update_request: I/O error, dev vda, sector 3799040 op 0x1:(WRITE) flags 0x4000 phys_seg 20 prio class 0
```

It also regenerates if I do the following inside the VM:

```
$ dd if=/dev/random of=1.txt bs=512 count=1000000
$ sync 1.txt
sync: error syncing '1.txt': Input/output error
```

### What's wrong?

`iov_base=0x8000000000000000` in above `io_submit` is a memory pointer & seems to be an invalid memory address.

### Things I have tried

* Since this seems to be a memory issue, I tried increasing `--mem-size` of the spdk target to 4G. It didn't make a difference.
* When this error happened, we ran SPDK under unprivileged user. I tried running under `root`. It didn't make a difference & error happened again.
* Originally, we ran SPDK using systemctl. I tried running directly. It didn't make a difference & error happened again.

**Comment by pykello** [comment]
## Reproducing with QEMU
### 0. Setup
```
sudo apt install libguestfs-tools

export SPDK_HOME=/path/to/spdk

export KEY1="a033778e038b4679a506927e2681164f2a97c1a6967b09678d723d232243515f"

export KEY2="bff2c655cd4d100bd7388decba64a85ee227ecd86bc0c8941b3a8b30fd4f4043"
```

### 1. Download & encrypt image
```
wget https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-amd64.img

# set root password to cloud123
virt-customize -a bionic-server-cloudimg-amd64.img --root-password password:cloud123

# convert image to raw format
qemu-img convert -p -f qcow2 -O raw jammy-server-cloudimg-amd64.img jammy.raw

# convert to encrypted format with spdk_dd
echo '{"subsystems":[{"subsystem":"accel","config":[{"method":"accel_crypto_key_create","params":{"name":"super_key","cipher":"AES_XTS","key":"$KEY1","key2":"$KEY2"}}]},{"subsystem":"bdev","config":[{"method":"bdev_aio_create","params":{"name":"aio0","block_size":512,"filename":"spdk-encrypted-image","readonly":false}},{"method":"bdev_crypto_create","params":{"base_bdev_name":"aio0","name":"crypt0","key_name":"super_key"}}]}]}' \
| $SPDK_HOME/build/bin/spdk_dd --if jammy.raw --ob crypt0
```

**Comment by pykello** [comment]
## Repro attempt with NBD

DIDN'T REPRODUCE! It worked fine!

```
modprobe nbd

export SPDK_HOME=/opt/spdk

echo '{"subsystems":[{"subsystem":"accel","config":[{"method":"accel_crypto_key_create","params":{"name":"super_key","cipher":"AES_XTS","key":"a033778e038b4679a506927e2681164f2a97c1a6967b09678d723d232243515f","key2":"bff2c655cd4d100bd7388decba64a85ee227ecd86bc0c8941b3a8b30fd4f4043"}}]},{"subsystem":"bdev","config":[{"method":"bdev_aio_create","params":{"name":"aio0","block_size":512,"filename":"spdk-encrypted-image","readonly":false}},{"method":"bdev_crypto_create","params":{"base_bdev_name":"aio0","name":"crypt0","key_name":"super_key"}}]}]}' \
> spdk_conf.json

wget https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-amd64.img

# convert image to raw format
qemu-img convert -p -f qcow2 -O raw jammy-server-cloudimg-amd64.img jammy.raw

# convert to encrypted format with spdk_dd
touch spdk-encrypted-image
truncate -s 5G spdk-encrypted-image
$SPDK_HOME/bin/spdk_dd --config spdk_conf.json --if jammy.raw --ob crypt0

$SPDK_HOME/bin/vhost --config spdk_conf.json -S /var/tmp

$SPDK_HOME/scripts/rpc.py nbd_start_disk crypt0 /dev/nbd1

mkdir nbd_disk
mount /dev/nbd1p1 nbd_disk

cd nbd_disk/root
dd if=/dev/random of=1.txt bs=5 count=100000000
```

**Comment by ozgune** [comment]
Hey @pykello - quick question. Has this issue been fixed with #372?
